### PR TITLE
GeGLU FFN activation — cross-dataset TF/TFP/AF/DM

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

GeGLU (Dauphin et al. 2017, Shazeer 2020) replaces the standard two-matrix FFN with a gated variant:

```
FFN_GeGLU(x) = (x W_1) * GELU(x W_3) * W_2
```

where W_1, W_3 are up-projection matrices and W_2 is the down-projection. The GELU gate smoothly suppresses small activations near zero — a property directly useful in CFD surrogate models, where many mesh nodes have near-zero physical quantities (far-field pressure, low-velocity regions).

This is distinct from SwiGLU tested in PR #2938 (nezuko), which uses SiLU as the gate function. GELU has heavier tails and is smoother near zero compared to SiLU. In CFD regimes where node features span many decades, GELU's behavior at small magnitudes may better preserve gradient signal for low-intensity physics regions.

Key properties:
- GELU gate is differentiable everywhere, including near zero
- Heavier tails than SiLU: less aggressive suppression of mid-range activations
- Standard in many modern transformers (GPT-J, Falcon, CodeGen)
- No change to model depth or attention — pure FFN swap

Papers:
- Dauphin et al., Language Modeling with Gated Convolutional Networks (2017): https://arxiv.org/abs/1612.08083
- Shazeer, GLU Variants Improve Transformer (2020): https://arxiv.org/abs/2002.05202

## Instructions

Add a `--geglu` flag to `train.py`. When enabled, replace the FFN activation in all transformer blocks with GeGLU.

Implementation pattern:

```python
import torch
import torch.nn as nn
import torch.nn.functional as F

class GeGLUFFN(nn.Module):
    """
    GeGLU feed-forward block.
    FFN(x) = (x W_1) * GELU(x W_3) @ W_2
    Shazeer 2020: https://arxiv.org/abs/2002.05202
    """
    def __init__(self, hidden_dim: int, ffn_dim: int, dropout: float = 0.0):
        super().__init__()
        # Two parallel up-projections: one for value, one for gate
        self.W_gate = nn.Linear(hidden_dim, ffn_dim, bias=False)
        self.W_up   = nn.Linear(hidden_dim, ffn_dim, bias=False)
        self.W_down = nn.Linear(ffn_dim, hidden_dim, bias=False)
        self.dropout = nn.Dropout(dropout) if dropout > 0.0 else nn.Identity()

    def forward(self, x):
        gate = F.gelu(self.W_gate(x))
        up   = self.W_up(x)
        return self.dropout(self.W_down(gate * up))
```

Replace the standard FFN block in every transformer layer with `GeGLUFFN`. Keep `ffn_dim` consistent with the existing FFN hidden dimension (typically `4 * hidden_dim`). The parameter count increases modestly due to the extra W_gate matrix.

**Run all four datasets in a single W&B group:**

```bash
# TandemFoil (Lion, EMA=0.999, 3L/192d)
cd target/icml2026 && python train.py \
  --dataset tandemfoil \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 999 \
  --geglu \
  --wandb_group wave11-geglu

# TandemFoil Paper (Lion, EMA=0.999, same arch)
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 --cosine-t-max 10 \
  --grad-clip 0.5 --weight-decay 1e-2 \
  --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction --enable-pressure-prior-addition \
  --ema-decay 0.999 \
  --epochs 999 \
  --geglu \
  --wandb_group wave11-geglu

# AirfRANS (AdamW, no-EMA, 2L/256d)
cd target/icml2026 && python train.py \
  --dataset airfrans --airfrans-task full \
  --optimizer adamw --lr 6e-4 --cosine-t-max 50 \
  --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema \
  --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 \
  --geglu \
  --wandb_group wave11-geglu

# DrivAerML (AdamW, no-EMA, 4L/512d)
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py \
  --dataset drivaerml \
  --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema \
  --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 394 --max-eval-batches 200 \
  --geglu \
  --wandb_group wave11-geglu
```

Run all four in parallel across your 8 GPUs. Report W&B run IDs for all four runs.

**Note on parameter count:** GeGLU adds one extra up-projection matrix per FFN (W_gate). If you want to maintain strict parameter parity with the baseline, reduce ffn_dim to approximately `ffn_dim * 2/3` (as recommended by Shazeer). This is optional — try the default ffn_dim first and note whether increased capacity helps or hurts.

## Baseline

Current best metrics (from BASELINE.md):

| Dataset | Primary Metric | Value | PR | W&B Run |
|---------|---------------|-------|-----|---------|
| TandemFoil | `val_primary/surface_pressure_mae` | **22.537** | #2924 | 0lv7fnun |
| AirfRANS | `val_primary/surface_mse` | **0.000482** | #2951 | pr4wsbfm |
| DrivAerML | `val_primary/surface_rel_l2_pct` | **3.997%** | #2898 | bht6h42t |

**TandemFoil reproduce command:**
```bash
cd target/icml2026 && python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 999 --ema-decay 0.999
```

**AirfRANS reproduce command:**
```bash
cd target/icml2026 && python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999
```

**DrivAerML reproduce command:**
```bash
cd target/icml2026 && SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200
```

## Relationship to SwiGLU (PR #2938)

PR #2938 (nezuko) tested SwiGLU (`FFN(x) = (xW_1) * SiLU(xW_3) @ W_2`) and is currently WIP. This PR tests GeGLU independently. The two hypotheses differ in gate activation: SiLU vs GELU. If both beat baseline, the results should be compared directly to pick the better gate function. If only one wins, the winning gate type becomes part of the standard recipe.

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT casca:

## Results

### GeGLU FFN Activation — Cross-Dataset Results

| Dataset | Primary Metric | GeGLU Best Val | Baseline | Δ | GeGLU Best Test | Epochs | Best Epoch | W&B Run |
|---------|---------------|----------------|----------|---|-----------------|--------|------------|---------|
| TandemFoil | `surface_pressure_mae` | **23.265** | 22.537 | +3.2% | 24.361 | 292 | 290 | `7u3lumdq` |
| TandemFoil Paper | `field_mse` | **NaN** | — | — | NaN | 490 | 1 | `e7fm7vp7` |
| AirfRANS | `surface_mse` | **0.000800** | 0.000482 | +66% | 0.001368 | 678 | 301 | `c04kt42i` |
| DrivAerML | `surface_rel_l2_pct` | **10.302%** | 3.997% | +158% | 10.980% | 470 | 49 | `b5bfvzy6` |

All experiments ran to the 360-minute timeout. None beat baseline.

### Commands Used

```bash
# TandemFoil (GPU 2)
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 --geglu --agent casca --wandb-name "casca/geglu-tandemfoil" --wandb-group wave11-geglu

# TandemFoil Paper (GPU 3)
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --model-slices 64 --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-fourier --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --ema-decay 0.999 --epochs 999 --geglu --agent casca --wandb-name "casca/geglu-tandemfoil-paper" --wandb-group wave11-geglu

# AirfRANS (GPU 4)
CUDA_VISIBLE_DEVICES=4 python train.py --dataset airfrans --airfrans-task full --optimizer adamw --lr 6e-4 --cosine-t-max 50 --grad-clip 1.0 --weight-decay 1e-2 --no-use-ema --enable-fourier --model-layers 2 --model-hidden-dim 256 --model-heads 4 --epochs 999 --geglu --agent casca --wandb-name "casca/geglu-airfrans" --wandb-group wave11-geglu

# DrivAerML (GPU 5)
CUDA_VISIBLE_DEVICES=5 SENPAI_MAX_EPOCHS=9999 python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 512 --model-heads 8 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 394 --max-eval-batches 200 --geglu --agent casca --wandb-name "casca/geglu-drivaerml" --wandb-group wave11-geglu
```

### Peak GPU Memory

| Dataset | GPU | Peak VRAM |
|---------|-----|-----------|
| TandemFoil | GPU 2 | ~46.8 GiB |
| TandemFoil Paper | GPU 3 | ~69.5 GiB |
| AirfRANS | GPU 4 | ~67.3 GiB |
| DrivAerML | GPU 5 | not captured at peak |

### What happened

**GeGLU is a clear negative result across all 4 datasets.** No dataset beat its baseline.

**TandemFoil (closest to baseline, +3.2%):** Converged slowly (~1.7 min/epoch due to torch.compile JIT overhead with the new module) but kept improving through all 292 epochs. Best val 23.265 vs baseline 22.537. Early epochs showed ~27 MAE, and the model was still improving when the 360-min timeout hit — suggesting GeGLU doesn't fundamentally break TF but learns more slowly. Longer training might close the gap but is unlikely to beat baseline.

**TandemFoil Paper (complete NaN):** Pressure channel produced NaN from epoch 1 onward. Ux and Uy channels learned normally, but the pressure channel diverged immediately. This is a GeGLU + Lion optimizer interaction specific to TFP's pressure distribution — the gated pathway appears to amplify numerical instability in the pressure channel.

**AirfRANS (converged but +66% worse):** Best val 0.000800 surface MSE vs baseline 0.000482 at epoch 301. The model then became unstable — final test MSE of 0.595 shows it diverged badly after the best checkpoint. The cosine schedule (T_max=50) means epochs 301+ were past several annealing cycles, and the gated FFN apparently makes the model more sensitive to LR cycling.

**DrivAerML (catastrophic instability, +158%):** Trained normally through epoch 49 (best val 10.302%), then entered catastrophic instability. Grad norms escalated from ~0.7 to 273+ by epoch 136, entering permanent divergence. Best EMA checkpoint preserved at 10.3% but 2.6x worse than the 3.997% baseline.

### Analysis

The hypothesis that GELU's heavier tails and smoother near-zero behavior would benefit CFD surrogates did not hold:

1. **Gating destabilizes training:** 2 of 4 datasets showed catastrophic instability (TFP from epoch 1, DM from epoch 57). The gating mechanism creates sharper loss landscape features that the current optimizer/LR configurations can't navigate.
2. **Parameter inefficiency:** GeGLU adds ~50% more parameters per FFN block (extra W_gate matrix) without ffn_dim reduction. The increased capacity doesn't translate to better metrics — the extra parameters may contribute to overfitting or optimization difficulty.
3. **Lion incompatibility:** Lion optimizer appears fundamentally incompatible with GeGLU on TFP. AdamW datasets avoided NaN but still showed instability. This suggests the gated FFN's gradient statistics differ enough from standard FFN that optimizer tuning is needed.
4. **Comparison with SwiGLU (PR #2938):** Independent from SwiGLU which uses SiLU as the gate. If SwiGLU also fails, it would suggest gated FFNs in general don't suit the Transolver architecture at current scale.

### Suggested follow-ups

1. **Reduced ffn_dim (2/3 factor):** The PR mentioned this for parameter parity. Worth testing to isolate whether the extra parameters hurt.
2. **Lower learning rate:** GeGLU's gating dynamics may require 0.5x current LR to avoid instability.
3. **AdamW for TF/TFP instead of Lion:** The Lion+GeGLU combination is clearly problematic.
4. **Tighter grad clipping for DM:** Clipping at 0.1 instead of default might prevent the catastrophic divergence.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3017 because ADVISOR: Closing — GeGLU is a clear negative result across all four datasets.

**Results (verified against W&B):**
| Dataset | Best Val | Baseline | vs Baseline | W&B |
|---|---|---|---|---|
| TandemFoil | 23.265 (ep290) | 22.537 | +3.2% worse | 7u3lumdq |
| TandemFoil Paper | NaN | 0.00434 | failed | e7fm7vp7 |
| AirfRANS | 0.000800 (ep301) | 0.000482 | +66% worse | c04kt42i |
| DrivAerML | 10.302% (ep49) | 3.997% | +158% worse | b5bfvzy6 |

The gating mechanism destabilizes training on complex 3D datasets. DM grad norms escalated from ~0.7 to 273+ before diverging. TFP NaN from epoch 1 (Lion+GeGLU incompatibility on pressure channel). AF diverged post-peak despite reaching a reasonable transient basin.

TF's near-miss (23.265, only +3.2%) suggests gated FFNs can converge on simpler 2D geometry but scale poorly to 3D CFD. We'll wait for SwiGLU (#2954/nezuko) to decide if the entire GLU-variant family is closed.

Good analytical work in the results — your identification of the Lion+GeGLU incompatibility and the grad norm escalation pattern is valuable..
